server/src/main/java/org/elasticsearch/bootstrap/Security.java

```diff
@@ -221,14 +221,16 @@ static Policy readPolicy(URL policyFile, Map<String, URL> codebases) {
                     if (aliasProperty.equals(property) == false) {
                         propertiesSet.add(aliasProperty);
                         String previous = System.setProperty(aliasProperty, url.toString());
-                        if (previous != null) {
+                        // allow to re-set to what was previously there. https://github.com/elastic/elasticsearch/issues/31324
```

Comment:
When I pulled this locally and reverted the changes to this file I didn't have any trouble. We've traditionally been very weary of making changes to this file so I'd really like to make sure we need this before we do it, even if it is temporary.

Reply:
Oh, wait, I just got it! Neat.

Reply:
This looks like it is caused by gradle not passing *any* system properties to the test sometimes. We use the absence of a system property set by gradle to engage "IDE mode" when running tests which mucks with the security features a bit. If you run tests from gradle in IDE mode they won't work right. Security, at least, won't work right. If these were to happen for other tests then it'd explode in other ways.

Reply:
I've dug some more. This is caused by us running the tests with the built in gradle test runner rather than the randomized runner. We configure the randomized runner to run with the system properties but we don't ever configure the standard runner.

Reply:
And some more: this is not caused by the build compare plugin. Maybe by gradle 4.8 or maybe by one of our hacks to make 4.8 work.

Reply:
Ok - I reverted the 4.8 change locally but kept the other changes and everything *looks* to work. I don't think we can upgrade to 4.8 until we get to the bottom of this.

Reply:
Thanks for looking into this @nik9000. I'll create a feature branch. 
I won't revert this change just yet, as  it does allow us to uncover additional issues, like the CI run here just found, the build hangs, and it's easily reproducible with  `./gradlew :example-plugins:painless-whitelist:check`. 
This upgrade of Gradle and JDK is turning out to be much harder than it should. 

I was under the impression that we always replace the Gradle test runner with the randomized one. I wonder if for some reason that doesn't happen, or doesn't always happen for some reason. I did have a suspicion that we are not running with the randomized runner, but when I looked at it I didn't find evidence of this being the case.